docs/blog/2021-06-16-tailor-your-slack-audit-sink-messages.md

```diff
@@ -0,0 +1,149 @@
+
+Modifications made to the audit event schema:
+
+```protobuf
+message RequestEvent {
+  ...
+  // The API request saved as metadata for the event.
+  RequestMetadata request_metadata = 7;
+  // The API response saved as metadata for the event.
+  ResponseMetadata response_metadata = 8;
+}
+
+message RequestMetadata {
+  google.protobuf.Any body = 1;
+}
+
+message ResponseMetadata {
+  google.protobuf.Any body = 1;
+}
+```
+
+The event’s Any fields can then be unpacked into its underlying message type by the audit sinks.
+
+Below is an example of an audit event with the stored metadata for [DescribePod](https://github.com/lyft/clutch/blob/a07d897dc69229ad1453fea6dcaf36ad50d2f6a6/api/k8s/v1/k8s.proto#L16-L22):
+
+```json
+{
+...
+"requestMetadata":{"body":{"@type":"type.googleapis.com/clutch.k8s.v1.DescribePodRequest","clientset":"kind-clutch-local","cluster":"kind-clutch-local","namespace":"envoy-staging","name":"envoy-58d647f457-5b2k4"}},
+"responseMetadata":{"body":{"@type":"type.googleapis.com/clutch.k8s.v1.DescribePodResponse","pod":{"cluster":"kind-clutch-local","namespace":"envoy-staging","name":"envoy-58d647f457-5b2k4","containers":[{"name":"envoy","image":"docker.io/envoyproxy/envoy:v1.14-latest","state":"RUNNING","ready":true,"restartCount":65,"stateRunning":{}}],"nodeIp":"172.18.0.2","podIp":"10.244.0.5","state":"RUNNING","labels":{"app":"envoy","bar":"value","env":"newstaging","foo":"value","pod-template-hash":"58d647f457","yett":"yes"},"podConditions":[{"type":"INITIALIZED","status":"TRUE"},{"type":"READY","status":"TRUE"},{"type":"CONTAINERS_READY","status":"TRUE"},{"type":"POD_SCHEDULED","status":"TRUE"}],"status":"Running","startTimeMillis":"1616607169000"}}}
+...
+}
+```
+
+## Richer Slack Audits
+
+With new metadata on hand in the audit event, we could support customized Slack messages. Our goals for the feature were to 1) avoid code duplication and 2) make it easy for users to set up for their workflows. Let’s dive into the latter point first.
+
+We [updated](https://github.com/lyft/clutch/blob/c50f64a6a487f7a961a333251afe2cdede3addb7/api/config/service/auditsink/slack/v1/slack.proto#L18-L43) the Slack configuration to allow users to specify a custom message for a `service/method` in the [clutch-config](https://github.com/lyft/clutch/blob/main/backend/clutch-config.yaml). The feature is powered by the Golang [template](https://golang.org/pkg/text/template/) package, so in the freeform message field you’d provide the names of the fields from the workflow’s API request/response. These fields will be replaced by their values when parsed by the Slack sink. We intentionally chose to append the custom message to the default Slack message for a richer Slack audit to prevent users from removing basic auditing with a bad override.
```

Comment:
```suggestion
We [updated](https://github.com/lyft/clutch/blob/c50f64a6a487f7a961a333251afe2cdede3addb7/api/config/service/auditsink/slack/v1/slack.proto#L18-L43) the Slack configuration to allow users to specify a custom message for a `service/method` in the [clutch-config](https://github.com/lyft/clutch/blob/main/backend/clutch-config.yaml). The feature is powered by the Golang [template](https://golang.org/pkg/text/template/) package, so in the freeform message field you’d provide the names of the fields from the workflow’s API request/response. These fields will be replaced by their values when parsed by the Slack sink. We intentionally chose to append the custom message to the standard Slack message to prevent users from removing basic auditing with a less detailed override.
```